pkg/storage/sqlstorage/balances.go

```diff
@@ -0,0 +1,202 @@
+	}
+
+	sb.Limit(int(q.Limit + 1))
+	t.Limit = q.Limit
+	sb.Offset(int(q.Offset))
+
+	balanceQuery, args := sb.BuildWithFlavor(s.schema.Flavor())
+	rows, err := exec.QueryContext(ctx, balanceQuery, args...)
+	if err != nil {
+		return sharedapi.Cursor[core.AccountsBalances]{}, s.error(err)
+	}
+
+	defer func(rows *sql.Rows) {
+		err := rows.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(rows)
+
+	accounts := make([]core.AccountsBalances, 0)
+
+	for rows.Next() {
+		var (
+			currentAccount string
+
+			assetBalance []byte
+		)
+		err = rows.Scan(&currentAccount, &assetBalance)
+
+		// we (clean and) marshal the computed row in a map[string]int64
+		assetBalances := arrayToAssetBalance(assetBalance)
+		if err != nil {
+			return sharedapi.Cursor[core.AccountsBalances]{}, s.error(err)
+		}
+
+		accounts = append(accounts, map[string]map[string]int64{
+			currentAccount: assetBalances,
+		})
+	}
+
+	if err := rows.Err(); err != nil {
+		return sharedapi.Cursor[core.AccountsBalances]{}, s.error(err)
+	}
+
+	var previous, next string
+	if q.Offset-q.Limit > 0 {
+		t.Offset = q.Offset - q.Limit
+		raw, err := json.Marshal(t)
+		if err != nil {
+			return sharedapi.Cursor[core.AccountsBalances]{}, s.error(err)
+		}
+		previous = base64.RawURLEncoding.EncodeToString(raw)
+	}
+
+	if len(accounts) == int(q.Limit+1) {
+		accounts = accounts[:len(accounts)-1]
+		t.Offset = q.Offset + q.Limit
+		raw, err := json.Marshal(t)
+		if err != nil {
+			return sharedapi.Cursor[core.AccountsBalances]{}, s.error(err)
+		}
+		next = base64.RawURLEncoding.EncodeToString(raw)
+	}
+
+	return sharedapi.Cursor[core.AccountsBalances]{
+		PageSize: len(accounts),
+		Previous: previous,
+		Next:     next,
+		Data:     accounts,
+	}, nil
+}
+
+func (s *Store) GetAggregatedBalancesData(ctx context.Context, exec executor, q storage.BalancesQuery) (map[string]int64, error) {
+	sb := sqlbuilder.NewSelectBuilder()
+	sb.Select("asset", "sum(input - output)")
+	sb.From(s.schema.Table("volumes"))
+	sb.GroupBy("asset")
+
+	if q.Filters.Address != "" {
+		arg := sb.Args.Add("^" + q.Filters.Address + "$")
+		switch s.Schema().Flavor() {
+		case sqlbuilder.PostgreSQL:
+			sb.Where("account ~* " + arg)
+		case sqlbuilder.SQLite:
+			sb.Where("account REGEXP " + arg)
+		}
+	}
+
+	balanceAggregatedQuery, args := sb.BuildWithFlavor(s.schema.Flavor())
+
+	rows, err := exec.QueryContext(ctx, balanceAggregatedQuery, args...)
+	if err != nil {
+		return nil, s.error(err)
+	}
+
+	defer func(rows *sql.Rows) {
+		err := rows.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(rows)
+
+	aggregatedBalances := make(map[string]int64)
+
+	for rows.Next() {
+		var (
+			asset  string
+			amount int64
+		)
+		err = rows.Scan(&asset, &amount)
+		if err != nil {
+			return nil, s.error(err)
+		}
+
+		aggregatedBalances[asset] = amount
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, s.error(err)
+	}
+
+	return aggregatedBalances, nil
+}
+
+func arrayToAssetBalance(array []byte) map[string]int64 {
```

Comment:
You could implements the sql.Scanner interface.

Reply:
Yes it's true ! I tried (right now) and even though it's a more elegant solution, i couldnt manage to do a proper unit test for this method, so i let the function as it is right now. If you still think it's better with the scanner interface, it's ok for me but i'd love to get some input about how to do a clean unit test for this method.

(i've shared my implementation with @antoinegelloz)